query_server/query/src/extension/physical/plan_node/update_tag.rs

```diff
@@ -0,0 +1,211 @@
+            assigns: self.assigns.clone(),
+            coord: self.coord.clone(),
+        }))
+    }
+
+    fn execute(
+        &self,
+        partition: usize,
+        context: Arc<TaskContext>,
+    ) -> Result<SendableRecordBatchStream> {
+        let scan = self.scan.execute(partition, context)?;
+
+        Ok(Box::pin(RecordBatchStreamAdapter::new(
+            self.schema(),
+            futures::stream::once(do_update(
+                self.schema.clone(),
+                self.table_schema.clone(),
+                scan,
+                self.assigns.clone(),
+                self.coord.clone(),
+            ))
+            .try_flatten(),
+        )))
+    }
+
+    fn metrics(&self) -> Option<MetricsSet> {
+        Some(self.metrics.clone_inner())
+    }
+
+    fn fmt_as(&self, t: DisplayFormatType, f: &mut std::fmt::Formatter) -> std::fmt::Result {
+        match t {
+            DisplayFormatType::Default => {
+                write!(f, "UpdateValueTagExec: [{}]", self.table_schema.name,)
+            }
+            DisplayFormatType::Verbose => {
+                let schemas = self
+                    .schema
+                    .fields()
+                    .iter()
+                    .map(|e| e.name().to_string())
+                    .collect::<Vec<_>>();
+                write!(
+                    f,
+                    "UpdateValueTagExec: [{}], output=[{}]",
+                    self.table_schema.name,
+                    schemas.join(",")
+                )
+            }
+        }
+    }
+
+    fn statistics(&self) -> Statistics {
+        self.scan.statistics()
+    }
+}
+
+async fn do_update(
+    schema: SchemaRef,
+    table_schema: TskvTableSchemaRef,
+    scan: SendableRecordBatchStream,
+    assigns: Vec<(String, Arc<dyn PhysicalExpr>)>,
+    coord: CoordinatorRef,
+) -> Result<SendableRecordBatchStream> {
+    let mut new_tags = vec![];
+    for (name, expr) in assigns {
+        let value = expr
+            .evaluate(&RecordBatch::new_empty(Arc::new(Schema::empty())))?
+            .into_array(1);
+        let value = value
+            .as_any()
+            .downcast_ref::<StringArray>()
+            .ok_or(DataFusionError::Execution(String::from(
+                "tag value must be string",
+            )))?
+            .value(0);
+        let update_set_value = UpdateSetValue {
+            key: name.as_bytes().to_vec(),
+            value: Some(value.as_bytes().to_vec()),
+        };
+        new_tags.push(update_set_value);
+    }
+
+    let mut rows_writed = 0;
+    let mut stream: DropEmptyRecordBatchStream = DropEmptyRecordBatchStream::new(scan);
+    while let Some(batch) = stream.next().await {
+        let batch = batch?;
+        let num_rows = batch.num_rows();
+        rows_writed += num_rows;
+        coord
+            .update_tags_value(table_schema.clone(), new_tags.clone(), batch)
+            .await
+            .map_err(|err| DataFusionError::Execution(err.to_string()))?;
+    }
```

Comment:
It is best to find all the matched series at once and send only one request to each kv node, because this involves the problem of failed retries.

Reply:
ok, i refactor it

Reply:
ok